eyelid-match-ops/src/primitives/poly.rs

```diff
@@ -136,3 +157,87 @@ pub fn poly_split(a: &Poly) -> (Poly, Poly) {
 
     (al, ar)
 }
+
+/// Returns the inverse in the cyclotomic ring, if it exists.
+/// Otherwise, returns an error.
+pub fn inverse(a: &Poly) -> Result<Poly, String> {
+    let mut mod_pol = Poly::zero();
+
+    mod_pol[MAX_POLY_DEGREE] = Coeff::one();
+    mod_pol[0] = Coeff::one();
+
+    let y = extended_gcd(&mod_pol, a);
+    let mul_both = a.clone().mul(y.clone());
+    if mul_both.is_one() {
+        Ok(y)
+    } else {
+        Err("No inverse in the ring.".to_owned())
+    }
+}
+
+/// Helps to calculate the equation `cur = prev - q.cur`.
+fn update_diophantine(prev: Poly, cur: Poly, q: Poly) -> (Poly, Poly) {
+    let aux = cur.clone();
+    let mul_res = cur.mul(&q.clone());
+    let new_cur = prev.sub(&mul_res).clone();
+    let new_prev = aux.clone();
+
+    (new_cur, new_prev)
+}
+
+/// Returns polynomials x, y, d such that a.x + b.y = d.
+/// When d=1 we have that x is the multiplicative inverse of a.
+pub fn extended_gcd(a: &Poly, b: &Poly) -> Poly {
+    // Invariant a.xi + b.yi = ri
+
+    // init with x0=1, y0=0, r0=a
+    let mut x_prev = Poly::one();
+    let mut y_prev = Poly::zero();
+    let mut ri_prev = a.clone();
+    // next:     x1=0, y1=1, r1=b
+    let mut x_cur = Poly::zero();
+    let mut y_cur = Poly::one();
+    let ri_cur = b.clone();
+
+    let ri_aux = ri_cur.clone();
+    // TODO: q is just a monomial, then we can optimize the next computation
+    let (mut q, mut ri_cur) = DenseOrSparsePolynomial::from(ri_prev.clone())
+        .divide_with_q_and_r(&ri_cur.into())
+        .expect("init divisor is not zero");
```

Comment:
This will panic on `inverse(Poly::zero())`. Instead, return an error.

(See my detailed explanation below.)

Reply:
Did you need help fixing this?

Reply:
Rust has convenience methods and operators for this kind of result handling:
- `Option` to `Result`: https://doc.rust-lang.org/std/option/enum.Option.html#method.ok_or_else
- Only executing error code when needed: https://doc.rust-lang.org/book/ch13-01-closures.html
- Returning an error `Result`: https://doc.rust-lang.org/std/result/#the-question-mark-operator-

```suggestion
        .ok_or_else(|| "init divisor is not zero".to_string())?;
```

Reply:
It turns out this whole block of code is a duplicate of the code in the loop. After I removed the duplicate code, this change isn't needed any more.

(I also needed to implement a missing `Poly::divide_with_q_and_r()` method.)